### ISSUE TITLE
RFC: reshape-markdown

### ISSUE THREAD

**Issue by tmpfs** [reshape/reshape#32]
Hey!

I wanted to include markdown documents in my reshape templates so I created this:

https://github.com/makestatic/reshape-markdown

Usage is `markdown(src="document.md")` similar to an include file. Whilst the `section(md)` and `p(mdi)` markdown content helpers are useful I found my templates became littered with them and it was nicer to just dump a section of the markup in a markdown document and load it.

Could also be useful when you have existing content like a `README.md` file that you want to include in your reshape project.

**Comment by jescalan** [comment]
This is great, thanks for making this! How was your experience working on a plugin?

If you add the keyword `reshape-plugin` to your `package.json` and publish, it will show up automatically on http://plugins.reshape.ml within a day 😁

**Comment by tmpfs** [comment]
Thanks! It already has the keyword as it was derived from your `reshape-include` package :)

I think it would be useful to also be able to include raw html, in which case maybe I should rename it to `reshape-source`.

I am debating whether to add another plugin `reshape-source` which loads pre-rendered HTML files (useful when I have already converted my markdown to HTML - for example when I want to highlight code blocks using `source-highlight`) or whether I should just test the file extension and if it ends with `.html` bypass markdown processing. What do you think?

I would prefer to keep just the once plugin to do both jobs but then the `markdown` name is a misnomer hence the idea of renaming to `reshape-source`.

**Comment by tmpfs** [comment]
Overall the process of building the plugin was very easy, especially as it was derived from your existing plugin ;)

I would just ask that you remove the call to `open` in your `coverage` scripts because for those of us running Linux the `open` command doesn't work well (I use another command).

Maybe a script called `cover` that generates the coverage report and then `coverage` could just be `npm run cover && open [...]`.

Food for thought ;)

**Comment by jescalan** [comment]
So I think the ultimate solution here is a feature addition to the include plugin, which adds a webpack loader-type functionality which would allow different parsers to be used on included files. For example:

```js
plugins: [
  include({
    rules: [
      { test: /\.md$/, parser: MarkdownParser },
      { test: /\.html$/, parser: HtmlParser }
      { test: /\.svg$/, parser: SvgParser }
  })
]
```

This way you could run an include as usual and have it parsed by the correct parser and output correctly into the reshape tree. And best of all, it would require no extra syntax when writing tags, just for the configuration. And we could add common include types to reshape-standard (like the ones above) to make it extra easy for people.

**Comment by jescalan** [comment]
For your comment on `open`, that's totally fine, I would gladly accept that as a PR!

**Comment by tmpfs** [comment]
I agree that parsers based on pattern matching is the way forward, hopefully I'll get some time to do that soon then I can deprecate the `reshape-markdown` plugin :)

**Comment by jescalan** [comment]
@tmpfs Just made this functionality, will have it shipped shortly:

https://github.com/reshape/include/pull/17

It was actually remarkably few lines of code!

**Comment by tmpfs** [comment]
That is awesome! :) Let me know when it ships and I'll refactor `reshape-markdown` to an include parser.

**Comment by jescalan** [comment]
Ok, just shipped last week! 🙌

**Comment by tmpfs** [comment]
Great :) I just refactored and published a minor version update that uses `parserRules`, I'll try it out in the context of a real world application shortly. It removed a load of redundant code paths that are already being handled by the include plugin which is sweet!

I wondered a little about [this line](https://github.com/reshape/include/blob/master/lib/index.js#L38) in the include plugin:

```javascript
const includeAst = parser(src, { filename: includePath }, { filename: includePath })
```

Is it necessary to pass the same object twice?

**Comment by jescalan** [comment]
So it is actually necessary, but the reason appears to be because of a fluke with the way that I wrote the API for the sugarml parser. Check it out:

https://github.com/reshape/sugarml/blob/master/lib/index.js#L4

For some reason I appear to have been not thinking straight and made an extra `parseOptions` param that is completely unused as the second option, and the third is the actual options object that is used. However, here on the default parser:

https://github.com/reshape/parser/blob/master/lib/index.js#L5

The options are passed as the second param. So in order for that line to work with both the default parser and the sugarml parser, at the moment, it does need to be there twice.

That being said, there is no question that sugarml should have the second option removed haha -- if you'd like to PR this, since you caught it, you are more than welcome to do so. If not, I'll take care of it.

While I was poking around here, I figured I'd check what options can actually be passed to the parser. It turns out it's just the filename right now. This is the case for both sugarml and the default parser. While this is already documented for the default parser, I should probably add this to the docs for sugarml as well.
